Created generic virtualization article

gangwerz opened the pull request:
Additionally, this Virtualization article covers the use of GNOME Boxes, instead of VirtualBox. Boxes is simpler and more approachable for our customers. This article can then be expanded with Virtual Machine Manager (VMM) if customers want/need more advanced features. Boxes and VMM both use `libvirt` as the hypervisor backend, so Host/Guest machine compatibility would be the same.

gangwerz reviewed (commented):
Merged in the updates to master, and re-removed the `install-in-vm.md` file to clear the merge conflicts

jacobgkau commented:
Is there a reason the VirtualBox article needs to be removed instead of just adding a Boxes article or keeping VirtualBox as an option in the same article? I can think of the following reasons we'd want to keep it:

- Windows and macOS users wanting to try out Pop!_OS in a virtual machine wouldn't have GNOME Boxes (or VMM, for that matter) as an option.
- VirtualBox may have some extra brand recognition over GNOME Boxes (seems like it would be second only to VMware.)
- While GNOME Boxes may be easier to get up-and-running, VirtualBox makes it easier to do some slightly more advanced things (like custom networking setups.)

gangwerz commented:
The goal was to simplify the articles. We had an Install Pop in VirtualBox article, then a section of the Windows article was dedicated to setting up and running a VM on Pop to run Windows. 

I also plan to add a Virtual Machine Manager section, which would have the advanced features of Virtualbox, and more.

We wanted to move away from VirtualBox, as `libvirt` just works, and there are no DKMS modules needing to rebuild with kernel updates.

---

We could leave the VirtualBox for Pop article, as we do not have a guide for a Pop VM on Windows. But, we also do not ship Windows systems. There are a plethora of guides on running Linux VMs on Windows with a variety of Hypervisors for those so inclined.

I'm not against having a VBox for Pop article, but I'm trying to trim the fat a bit here. Do we think it is necessary for System76 to have a guide on running Pop on Windows? (genuine question, please read with no sarcasm)

jacobgkau commented:
> We had an Install Pop in VirtualBox article, then a section of the Windows article was dedicated to setting up and running a VM on Pop to run Windows.

Running Windows in a VM on Pop is useful for current users who need to run Windows software. Running Pop in a VM on Windows is useful for prospective users who want to try out the OS (persistently, and without risking damage to their current OS installation.) I see two very distinct use cases for these two articles.

> Do we think it is necessary for System76 to have a guide on running Pop on Windows?

I certainly think it's useful, yes. We want people to be able to try out Pop!_OS easily so they can see what it's capable of and experience our high standards of development. That's the same reason we put effort into supporting Pop!_OS users on non-System76 hardware, and that we ship the Launch keyboard configurator for Windows and macOS.

> There are a plethora of guides on running Linux VMs on Windows with a variety of Hypervisors for those so inclined.

Our first-party documentation should always provide a more confident experience for users than having to search the web for third-party documentation.

> I'm not against having a VBox for Pop article, but I'm trying to trim the fat a bit here.

I wasn't aware we were needing to "trim the fat." on the contrary, it seems like we've been adding a lot of new articles and trying to expand our documentation. I'm sure we can find the resources to maintain this article if that's the concern.

gangwerz commented:
I will re-add that install-in-vm article, as that is fine. What are your thoughts on making it a generic "Install Linux on Windows with VBox" article? That way we have a 1-size fits all for both the Windows and Linux hosts that look similar. I can leave `install-in-vm` alone as well, if we want it to be Pop specific.

The trimming the fat was more about unifying the VM information, and linking to the install guides for the respective service. I did not like having a Installing Windows in a VM buried in the Windows Install article. So the "Windows Install" had the fat I wanted to trim out, not the number of support articles.

jacobgkau commented:
> What are your thoughts on making it a generic "Install Linux on Windows with VBox" article?

I don't see why that needs to change. Like I just said, we create Pop!_OS and we want people to be able to use Pop!_OS. I don't understand why our article wouldn't talk about installing Pop!_OS specifically.

I would be all for generalizing the article from mentioning "your favorite Linux based OS" as the host to covering Windows, macOS, and Linux hosts. The steps are the same after getting VirtualBox installed. There's no need to call out Windows specifically in the article title.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+>
+> If either of these happen when you download, right-click the link above and select "Save Link As..." and save the file to your Downloads folder.
+
+## Virtualiaztion Features
```
looks like virtualization is misspelled here

storyaddict replied:
Fixed.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+tableOfContents: true
+---
+
+![GNOME Boxes with VMs](/images/gnome-boxes/boxes-full.png)
```
Minor aesthetics - this screenshot is massive with a lot of blank gray space. The text is difficult to read. Is it possible to crop?
![image](https://user-images.githubusercontent.com/80931038/159061534-49c695a5-9226-4f11-b47c-be2dbf50e0c0.png)

gangwerz replied:
The screenshot was to show what the Boxes interface looks like with VMs installed. There is no way to increase the font size in the app as far as I can see.

But, the text is just the names of operating systems, and is not important to the content.

jacobgkau replied:
Can you re-take the screenshot after resizing the window vertically so there's not so much empty space at the bottom? I would go about 1/3 of the current height, or leave enough room at the bottom that it's clear the list is over.

jacobgkau replied:
This is just a heads-up, but in Pop!_OS 22.04 pre-release, GNOME Boxes appears to be using libadwaita, as its interface has the Adwaita Dark color scheme instead of Pop Dark.

The screenshot doesn't quite look like Pop Dark, is it using a third-party theme?

gangwerz replied:
@Ampersandstorm I can certainly crop the photo, but I was wondering what your thoughts were with the above information. What do you think would work better for this article?

gangwerz replied:
Sorry, did not see @jacobgkau's messages when I responded. I'll redo the screen shots with the default Pop theming.

Ampersandstorm replied:
@gangwerz Thanks! Off-topic but I'm wondering what screenshot program you use to apply the slight shadow effect on the edges?

gangwerz replied:
@Ampersandstorm I was using the built-in GNOME Screenshot tool for all the pictures. I'm not sure if the shadow is because the photos are PNGs, or if it's due to CSS settings on this page

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+There are 2 drivers needed to get the full functionality out of your Linux VM with GNOME Boxes. The first is the `spice-vdagent` driver, and the other is the `spice-webdavd` driver. Both drivers are provided by the Open Source [Spice Project](https://www.spice-space.org/).
+
+Some Linux distributions include one or both of these drivers out of the box, while others will need them both installed. Additionally, your Linux distribution may not use the names used above, so you may need to reach out to those maintainers for more information. You can see an example of this discrepency below in the Arch Linux section. On Arch, the `spice-webdavd` package does not exist, but  ...
```
Minor clarity - This statement may be confusing since line 106 states the drivers are required. I would suggest combining these paragraphs: 

"Depending on the distribution, you may need to install additional drivers to gain video compatibility and file-sharing features. The [Spice Project](https://www.spice-space.org/) provides the following drivers: 

**spice-webdavd** - This driver allows sharing folders and files between the Host and VM.

**spice-vdagent**

- Allows the VM screen to resize automatically to match the GNOME boxes window
- Provides ability to copy/paste between the Host and VM
- Supports transferring files from client to agent (according to the [README.md])(https://gitlab.freedesktop.org/spice/linux/vd_agent)

>Note: The driver names may vary by distribution, as shown in the below Arch example: `phodav` is used instead of the `spice-webdavd` package name. We recommend referencing your specific distributions documentation for equivalent or alternative driver names.

gangwerz replied:
Really like the look of this rewrite! This should resolve the above 2 comments as well. I'll get this implemented when I add the new photos.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+![GNOME Boxes with VMs](/images/gnome-boxes/boxes-full.png)
+
+This is a quick tutorial on running virtual machines, or VMs, on Linux. Virtual machines allow you to run other operating systems from within your primary operating systems. This is a great way to run applications that require older software, run applications that require an OS other than Linux, or to try out other operating systems.
```
Minor conciseness - If you change the title to "Creating Virtual Machines on Linux" you can get rid of this first sentence. "Virtualization in Linux" can mean a few things.

gangwerz replied:
Fair point about the title, I'll update that :+1:

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+## Set Up
+
+First things first, we will need to a Linux operating system installed on your computer. We will be using Pop!_OS as the operating system running the virtualization software, called the Host operating system. Then, we will use Ubuntu as the "Guest" operating system, which runs inside the VM.
```
Minor scope - I would assume most people that come across this article will already be running Linux. This section is still useful, but it may be helpful to clarify with something like:

"This article assumes you are already running a Linux distribution. Instructions in this guide use Pop!_OS as the "Host" operating system: The operating system that runs the virtualization software. See instructions below if you aren't running Linux, or would like to install Pop!_OS to make things simple."

gangwerz replied:
I added a link to the Install Pop!_OS in a VM article to the links section, and made the suggested change.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+The install process for your chosen operating system will be just like any other install, with nothing VM specific at this time. Proceed through the OS install using the OS's guides, or possibly one of the System76 guides linked below:
+
+#### Linux
```
Minor clarity - Instead of just "Linux" change section header to "Linux Installation Steps".

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+After selecting the resources for your VM, the VM will boot the selected operating system installer. Likely, you will see some odd graphical behavior, such as the black bars shown in the above screenshot. This is because some of the drivers needed for the VM are not installed yet. These oddities can be ignored for now, and we will install the drivers later.
+
+The install process for your chosen operating system will be just like any other install, with nothing VM specific at this time. Proceed through the OS install using the OS's guides, or possibly one of the System76 guides linked below:
```
Minor brevity - consider:

"Complete the standard installation steps for your virtual operating system. These steps do not differ from a bare-metal installation shown in the below guides:"

Ampersandstorm replied:
I meant for this comment to only apply to line 84, the paragraph above is fine.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+When you attempt to create a Windows VM, you will likely see a prompt for the "Express Install" feature of GNOME Boxes. You can certainly use the Express Install with your Windows product key. However, for this article turn off the Express Install, then click "Next" at the top right to move to the standard installer set up.
+
+Install Windows - [https://support.system76.com/articles/windows/](https://support.system76.com/articles/windows/)
```
Minor clarity - It may be helpful to give this link some context (like you did for the Linux installers in line 84). It my seem obvious, but may confuse some readers if it's not explicitly stated that the link is offered as an optional reference. Consider this after the last sentence in line 96:

"Reference the below link for standard Windows installation steps."

gangwerz replied:
Reworked this section with the Linux installers and Express Install sections.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+- Windows Guest Tools: [https://www.spice-space.org/download/windows/spice-guest-tools/spice-guest-tools-latest.exe](https://www.spice-space.org/download/windows/spice-guest-tools/spice-guest-tools-latest.exe)
+- Spice WebDAVd: [https://www.spice-space.org/download/windows/spice-webdavd/spice-webdavd-x64-latest.msi](https://www.spice-space.org/download/windows/spice-webdavd/spice-webdavd-x64-latest.msi)
+
+> While trying to download the Spice WebDAVd file in the Edge browser, if I clicked the link, I was taken to a page trying to display the file as text instead of downloading the file. This ends up showing a lot of odd letters and characters.
```
Minor formatting & clarity - It's obvious this is offered to help users troubleshoot common pitfalls, but the sudden first-person is an odd convention compared to the rest of the article. Consider adding a "Troubleshooting" header, and/or converting these first-person scenarios to standard issue-resolution format: 

Issues: 

- Clicking the WebDavd file in an Edge Browser session takes the user to page that tries to display the file as jumbled text instead of downloading the file.
- After multiple download attempts, a user sees a window with random text or receives a "This page isn't responding" prompt. 

Resolution: Right click the link above and select "Save Link As..." and save teh file to your downloads folder.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+### Share a Folder
+
+_For this guide, we used GNOME-based Linux distributions, so the VMs UI closely matched Pop!_OS and Ubuntu. Accessing shared folders with other desktop environments is possible, but you will need to refer to their documentation for the details_
```
Minor formatting - this renders odd for me, not sure if this is just on my end or if this is a markdown syntax issue 
![image](https://user-images.githubusercontent.com/80931038/159072246-73bca338-8113-44ad-ada9-9ee213b420e4.png)

storyaddict replied:
I corrected this in my latest commit. :)

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+#### Selecting the Host's Folder
+
+In this pop-up, the "Local Folder" refers to the folder's name on the Host system. This defaults to the "Public" folder in your Home directory. If you click the drop down, you can quickly select some of the important folders on your system. If you do not wish to use one of these folders, select "Other..." This will open a folder selection window, where you can then navigate to the folder you wish ...
```
Minor formatting - This can stay as is for now, but going forward I've always preferred reading step-by-step instructions in step format. In paragraph form, this can be a lot for a reader to keep in their head, especially since there's no visual context. This breaks up the process, and gives the reader a reference point if they get stuck on a specific stage. Consider:

1. Click the drop down to quickly select important folders on your system.
2. If the folder you need isn't listed, click "Other" to navigate to the folder you want to share with the VM.
3. Once you have located the folder to share, click "Open" in the top right corner.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+![Accessing a Shared Folder](/images/gnome-boxes/boxes-vm-access-folder-share.png)
+
+Now that GNOME Boxes is sharing the folder, we can close the VM settings menu and move back to the VM. In the VM, open your Files application, then navigate to "Other Locations" on the left hand side. You will likely see a spinning loading wheel under "Networks" for a moment, so wait for that to finish. Once the loading is done, you should see "Spice client folder" listed under "Networks." Click  ...
```
Same as comment above. I realize this would be a lot to go back and change since the entire article is written this way, but just something to consider for future content.

Ampersandstorm reviewed (changes requested):
I like this article, it's pretty good right now but I've given some feedback to help readability and clarity.

storyaddict commented on content/virtualization.md:
```diff
@@ -0,0 +1,199 @@
+
+This driver brings in support for what most users will need. After installing this package and rebooting the VM, you will notice the VM will resize to match the size of your GNOME Boxes window. Additionally you will now be able to copy and paste between the Host and the VM. Both of these features are enabled by default once the driver is installed, so no configuration necessary.
+
+##### WebDAVd
```
I think my latest commit addresses this, at least in part.

gangwerz commented:
Finished review the change suggestions, and all make perfect sense to me. I will update the article with these in mind, and resolve the conversations as I make the changes.

gangwerz commented:
Ending my changes for the day.

- Update photos to show default appearance with adwaita theming
  - Shrunk window size to minimize whitespace
  - GNOME Boxes is a whitespacey app, so still a bit here and there.
- Incorporated @Ampersandstorm's suggestions
  - Made instructions beyond 2 steps into numbered lists. Left the "Creating a VM" section as paragraphs. Felt the photos broke the section up into pieces well enough. Adding a numbered list looked weird to me when I tested here.
  - Added the Drag and Drop functionality to the `vd-agent` section. I tested, and this worked in Fedora, Ubuntu, and AlmaLinux. So, seems to be pretty well supported at this point.
- Added a link to the Install Pop!_OS in a VM article, in case anyone opens the wrong article
- Reworked the "Install GNOME Boxes Drivers" section to flow better.
- Made the "Express Install" section generic. I saw this when testing some Linux distros, in addition to Windows. Not all Linux distros showed this feature however.

Ampersandstorm commented on content/virtualization.md:
```diff
@@ -0,0 +1,244 @@
+
+![Devices and Shares](/images/gnome-boxes/boxes-vmprefs-usb.png)
+
+One of the features of GNOME Boxes, is the ability to transfer a USB device connected to your Host to the VM. To share a USB device
```
missing colon after "To share a USB device"

jacobgkau replied:
There should not be a comma after "GNOME Boxes" in this sentence, either.

Ampersandstorm reviewed (changes requested):
Just found one missing colon, otherwise changes look great.

storyaddict reviewed (approved):
Changes all look good to me. Viewed live as well.
